cmd/k8s-operator/proxygroup_specs.go

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+package main
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	tsapi "tailscale.com/k8s-operator/apis/v1alpha1"
+	"tailscale.com/types/ptr"
+)
+
+const labelSecretType = "tailscale.com/secret-type"
+
+// Returns the base StatefulSet definition for a ProxyGroup. A ProxyClass may be
+// applied over the top after.
+func pgStatefulSet(pg *tsapi.ProxyGroup, namespace, cfgHash string) *appsv1.StatefulSet {
+	return &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            pg.Name,
+			Namespace:       namespace,
+			Labels:          labels("proxygroup", pg.Name, nil),
+			OwnerReferences: pgOwnerReference(pg),
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Replicas: ptr.To(pgReplicas(pg)),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels("proxygroup", pg.Name, nil),
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:                       pg.Name,
+					Namespace:                  namespace,
+					Labels:                     labels("proxygroup", pg.Name, nil),
+					DeletionGracePeriodSeconds: ptr.To[int64](10),
+					Annotations: map[string]string{
+						podAnnotationLastSetConfigFileHash: cfgHash,
+					},
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: pg.Name,
+					InitContainers: []corev1.Container{
+						{
+							Name:  "sysctler",
+							Image: fmt.Sprintf("tailscale/tailscale:%s", selfVersionImageTag()),
+							SecurityContext: &corev1.SecurityContext{
+								Privileged: ptr.To(true),
+							},
+							Command: []string{
+								"/bin/sh",
+								"-c",
+							},
+							Args: []string{
+								"sysctl -w net.ipv4.ip_forward=1 && if sysctl net.ipv6.conf.all.forwarding; then sysctl -w net.ipv6.conf.all.forwarding=1; fi",
```

Comment:
It would have been better to use the existing templates that we have https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/deploy/manifests/proxy.yaml
These proxies are meant to be configured in the same way as the ones from the templates- now we'll have the same config in two places (potential for drift, need to make changes in two places etc)

Reply:
Sorry if I misunderstood, I thought this was part of the clean break we discussed. That said, I would probably prefer to keep it this way so that the specs are defined somewhat completely and all together in the same file, as there's a non-trivial amount of linkage between them all.

My thinking with maintenance is that we won't need to make updates to the original proxies for long anyway, as we'd like to standardise on one workflow (i.e. default ProxyGroup) as soon as possible, but maybe I'm too optimistic about the timeline for that.

Reply:
That's fair, it is definitely a lot noisier visually to read the Go version than the yaml version. As discussed offline I'll note this down as a follow-up to save adding too much churn in this PR.